File: .tmp_c4000/batch_04_go/3481_apply_substitutions.go
// LeetCode 3481 - Apply Substitutions
// https://leetcode.com/problems/apply-substitutions/

func applySubstitutions(replacements [][]string, text string) string {
	mp := map[string]string{}
	for _, r := range replacements {
		mp[r[0]] = r[1]
	}
	var resolve func(string) string
	resolve = func(s string) string {
		out := []byte{}
		for i := 0; i < len(s); {
			if s[i] == '%' {
				j := i + 1
				for j < len(s) && s[j] != '%' {
					j++
				}
				key := s[i+1 : j]
				val := mp[key]
				out = append(out, resolve(val)...)
				i = j + 1
			} else {
				out = append(out, s[i])
				i++
			}
		}
		return string(out)
	}
	return resolve(text)
}

File: .tmp_rs4013/rust_check17/3481_apply_substitutions.rs
struct Solution;
// LeetCode 3481 - Apply Substitutions
// https://leetcode.com/problems/apply-substitutions/

use std::collections::HashMap;

impl Solution {
    pub fn apply_substitutions(replacements: Vec<Vec<String>>, text: String) -> String {
        let mut mp = HashMap::new();
        for r in replacements {
            mp.insert(r[0].clone(), r[1].clone());
        }
        fn resolve(s: &str, mp: &HashMap<String, String>) -> String {
            let bytes = s.as_bytes();
            let mut out = String::new();
            let mut i = 0;
            while i < bytes.len() {
                if bytes[i] == b'%' {
                    let mut j = i + 1;
                    while j < bytes.len() && bytes[j] != b'%' {
                        j += 1;
                    }
                    let key = &s[i + 1..j];
                    let val = mp.get(key).map(|v| v.as_str()).unwrap_or("");
                    out.push_str(&resolve(val, mp));
                    i = j + 1;
                } else {
                    out.push(bytes[i] as char);
                    i += 1;
                }
            }
            out
        }
        resolve(&text, &mp)
    }
}

fn main() {}
